ulib/FStar.Matrix.fst

```diff
@@ -0,0 +1,839 @@
+   into the concatenation of its first (m-1) rows and its last row (thus snoc)  *)
+let matrix_append_snoc_lemma #c (#m #n: pos) (generator: matrix_generator c m n)
+  : Lemma (matrix_seq generator == (SB.slice (matrix_seq generator) 0 ((m-1)*n))
+                                   `SB.append`
+                                   (SB.slice (matrix_seq generator) ((m-1)*n) (m*n))) 
+  = SB.lemma_eq_elim (matrix_seq generator) 
+                  (SB.append (SB.slice (matrix_seq generator) 0 ((m-1)*n))
+                          (SB.slice (matrix_seq generator) ((m-1)*n) (m*n)))
+
+let matrix_seq_decomposition_lemma  #c (#m:greater_than 1) (#n: pos) (generator: matrix_generator c m n)
+  : Lemma ((matrix_seq generator) == 
+          SB.append (matrix_seq #c #(m-1) #n generator)
+                    (SB.slice (matrix_seq generator) ((m-1)*n) (m*n)))
+  = SB.lemma_eq_elim (matrix_seq generator)
+                  ((matrix_seq #c #(m-1) #n generator) `SB.append` 
+                  (SB.slice (matrix_seq generator) ((m-1)*n) (m*n))) 
+
+(* This auxiliary lemma establishes the equality of the fold of the entire matrix
+   to the op of folds of (the submatrix of the first (m-1) rows) and (the last row). *) 
+let matrix_fold_snoc_lemma #c #eq 
+                           (#m: not_less_than 2) 
+                           (#n: pos) 
+                           (cm: CE.cm c eq) 
+                           (generator: matrix_generator c m n)
+  : Lemma (assert ((m-1)*n < m*n);
+            SP.foldm_snoc cm (matrix_seq generator) `eq.eq` 
+    cm.mult (SP.foldm_snoc cm (matrix_seq #c #(m-1) #n generator))
+            (SP.foldm_snoc cm (SB.slice (matrix_seq #c #m #n generator) ((m-1)*n) (m*n))))
+  = SB.lemma_eq_elim (matrix_seq generator)
+                  ((matrix_seq #c #(m-1) #n generator) `SB.append` 
+                  (SB.slice (matrix_seq generator) ((m-1)*n) (m*n)));    
+    SP.foldm_snoc_append cm (matrix_seq #c #(m-1) #n generator) 
+                         (SB.slice (matrix_seq generator) ((m-1)*n) (m*n)) 
+
+let foldm_snoc_last #c #eq (cm: CE.cm c eq) (s: SB.seq c{SB.length s > 0})
+  : Lemma (SP.foldm_snoc cm s == cm.mult (snd (SProp.un_snoc s)) 
+                                         (SP.foldm_snoc cm (fst (SProp.un_snoc s)))) 
+  = ()
+
+
+let matrix_submatrix_lemma #c (#m: not_less_than 2) (#n: pos)  
+                           (generator: matrix_generator c m n)
+  : Lemma ((matrix_seq generator) == (matrix_seq (fun (i:under(m-1)) (j:under n) -> generator i j) `SB.append` SB.init n (generator (m-1))))
+  = SB.lemma_eq_elim (matrix_seq (fun (i:under (m-1)) (j:under n) -> generator i j)) (matrix_seq #c #(m-1) #n generator);
+    SB.lemma_eq_elim (SB.slice (matrix_seq generator) ((m-1)*n) (m*n)) (SB.init n (generator (m-1)));
+    matrix_seq_decomposition_lemma generator
+
+(* This one could be probably written more efficiently -- 
+   but this implementation also works. *)
+#push-options "--ifuel 1 --fuel 0 --z3rlimit 130" 
```

Comment:
This proof is brittle and with such a high rlimit it will surely fail at some point again as F* and its libraries evolve. 

Before merging this, can you please find a way to make it more efficient?

Reply:
honestly I only put that big rlimit because the test mysteriously failed.

On my working machine, f* verifies this under z3rlimit of 10 without problem... could it be a sign of regression somewhere beyond the current F* release binary?

Reply:
Anyway, the fixed proof seems to be stable now